File: tests/test_mpi3rma_fortran/null_pointer.f90
      ! currently doesn't work with gcc

      subroutine test_mpi3rma_null_pointer ( comm )

#if INCLUDE_TYPE==1
      implicit none
      
      include 'fompif.h'
#endif
#if INCLUDE_TYPE==2
      use fompi
      
      implicit none
#endif

      integer, intent(in) :: comm

      integer :: res
      integer :: res_free
      character(20) :: create_flavor
      integer :: win
      integer :: j
      integer(kind=MPI_ADDRESS_KIND) :: win_size

      do j=0,1
        ! test if a NULL pointer as base is valid */
        if (j .eq. 0) then
          win_size = 0
          call foMPI_Win_create(null(), win_size, 4, MPI_INFO_NULL, comm, win, res);
          write(create_flavor,*) "create"
        else
          win_size = 0
          call foMPI_Win_allocate( win_size, 4, MPI_INFO_NULL, comm, null(), win, res)
          write(create_flavor,*) "allocate"
        endif
        call foMPI_Win_free( win, res_free )

        if ( (res .eq. MPI_SUCCESS) .and. (res_free .eq. MPI_SUCCESS) ) then
          write (*,*) trim(create_flavor), " - No Error"
        else
          write (*,*) "Error: Return code of MPI_Win_", trim(create_flavor), &
            " was ", res, " and of MPI_Win_free was ", res_free, &
            " instead of ", MPI_SUCCESS
        endif
      enddo

      end subroutine test_mpi3rma_null_pointer

File: tests/test_foMPI-NA/null_pointer.c
#include "fompi.h"

void test_mpi3rma_null_pointer(MPI_Comm comm) {

  int res;
  int res_free;
  char create_flavor[20];
  foMPI_Win win;
  int j;

  for ( j=0 ; j<2 ; j++ ) {
    /* test if a NULL pointer as base is valid */
    switch (j) {
      case 0: res = foMPI_Win_create(NULL, 0, 4, MPI_INFO_NULL, comm, &win);
              snprintf(create_flavor, 20, "create");
              break;
      case 1: res = foMPI_Win_allocate( 0, 4, MPI_INFO_NULL, comm, NULL, &win);
              snprintf(create_flavor, 20, "allocate");
              break;
    }
    res_free = foMPI_Win_free( &win );

    if ( (res == MPI_SUCCESS) && (res_free == MPI_SUCCESS) ) {
      printf("%s - No Error\n", create_flavor);
    } else {
      printf("Error: Return code of MPI_Win_%s was %i and of MPI_Win_free was %i instead of %i\n", create_flavor, res, res_free, MPI_SUCCESS);
    }
  }

  return;
}
